uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs

```diff
@@ -0,0 +1,102 @@
+
+#[no_mangle]
+pub extern "C" fn {{ obj.ffi_init_callback().name() }}(callback: uniffi::ForeignCallback) {
+    uniffi::set_foreign_callback(&{{ foreign_callback_holder }}, callback);
+}
+
+// Make an implementation which will shell out to the foreign language.
+#[derive(Debug)]
+struct {{ trait_impl }} {
+  handle: u64
+}
+
+// We need Send so we can stash the callbacks in objects we're sharing with foreign languages
+// i.e. in the handle map.
+// Prepared for our target trait to declare:
+//  `trait {{ trait_name }}: Send + std::fmt::Debug`
+unsafe impl Send for {{ trait_impl }} {}
+
+impl {{ trait_name }} for {{ trait_impl }} {
+    {%- for meth in obj.methods() %}
+
+    {#- Method declaration #}
+    fn {{ meth.name() -}}
+    ({% call rs::arg_list_decl_with_prefix("&self", meth) %})
+    {%- match meth.return_type() %}
+    {%- when Some with (return_type) %} -> {{ return_type|type_rs }}
+    {% else -%}
+    {%- endmatch -%} { 
+
+    {#- Method body #}
+        uniffi::deps::log::debug!("{{ obj.name() }}.{{ meth.name() }}");
+        let callback = uniffi::get_foreign_callback(&{{ foreign_callback_holder }}).unwrap();
+
+    {#- Packing args into a RustBuffer #}
+        {% if meth.arguments().len() == 0 -%}
+        let args_rbuf = uniffi::RustBuffer::new();
+        {% else -%}
+        let mut args_buf = Vec::new();
+        {% for arg in meth.arguments() -%}
+            {{ arg.name()|write_rs("&mut args_buf", arg.type_()) -}};
+        {% endfor -%}
+        let args_rbuf = uniffi::RustBuffer::from_vec(args_buf);
+        {% endif -%}
+
+    {#- Calling into foreign code. #}
+        let ret_rbuf = unsafe { callback(self.handle, {{ loop.index }}, args_rbuf) };
+
+    {#- Unpacking the RustBuffer to return to Rust #}
+        {% match meth.return_type() -%}
+        {% when Some with (return_type) -%}
+        let vec = ret_rbuf.destroy_into_vec();
+        let mut ret_buf = vec.as_slice();
+        let rval = {{ "&mut ret_buf"|read_rs(return_type) }};
+        rval
+        {%- else -%}
+        uniffi::RustBuffer::destroy(ret_rbuf);
+        {%- endmatch %}
+    }
+    {%- endfor %}
+}
+
+unsafe impl uniffi::ViaFfi for {{ trait_impl }} {
+    type FfiType = u64;
+
+    fn lower(self) -> Self::FfiType {
+        self.handle
+    }
+
+    fn try_lift(v: Self::FfiType) -> uniffi::deps::anyhow::Result<Self> {
+        Ok(Self { handle: v })
+    }
+
+    fn write<B: uniffi::deps::bytes::BufMut>(&self, buf: &mut B) {
+        buf.put_u64(self.handle);
+    }
+
+    fn try_read<B: uniffi::deps::bytes::Buf>(buf: &mut B) -> uniffi::deps::anyhow::Result<Self> {
+        uniffi::check_remaining(buf, 8)?;
+        <Self as uniffi::ViaFfi>::try_lift(buf.get_u64())
+    }
+}
+
+impl Drop for {{ trait_impl }} {
+    fn drop(&mut self) {
+        let callback = uniffi::get_foreign_callback(&{{ foreign_callback_holder }}).unwrap();
+        unsafe { callback(self.handle, 0, Default::default()) };
```

Comment:
using a named const for the `0` here might make it easier to understand what this does at a glance; IIUC it would be something like `const IDX_CALLBACK_FREE = 0` or similar.

Reply:
The fact that we implement `Drop` here makes me very suspicious of writing these into a buffer etc, because we could serialize it to a buffer, drop it, then deserialize into a fresh instance with a handle that's been dropped.